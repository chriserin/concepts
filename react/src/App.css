@import url("https://fonts.googleapis.com/css?family=Libre+Franklin:300,400,700");
html,
* {
  font-family: "Libre Franklin", sans-serif; }

.App {
  display: grid;
  grid-template-rows: repeat(auto-fit, minmax(40px, 1fr));
  grid-template-areas: "header" "nav" "concepts" "footer"; }

.App-header {
  background-color: rgba(46, 170, 212, 0.1);
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  align-content: baseline;
  color: white;
  padding: 0.8vh 1vw;
  grid-area: header;
  border-bottom: 1px solid #2389ac; }
  .App-header .hr-container {
    display: flex; }
  .App-header h2 {
    margin: 0;
    align-self: flex-end; }
  .App-header h2 a.hr {
    text-decoration: none;
    display: inline-block;
    padding: 0.4rem 0 0rem 1.6rem;
    text-transform: uppercase;
    color: #082736;
    letter-spacing: 0.1em;
    font-weight: 900;
    font-size: 60%;
    line-height: 2;
    color: #ae1f23;
    background: url(https://d15zqjc70bk603.cloudfront.net/assets/brand/hr_logo_icon-81c617c883074d68d44a316868516dc57b3c22eebb0dd3625da85895e554d7fb.svg) left 55% no-repeat;
    background-size: 1.2rem auto;
    font-family: Raleway, "Helvetica Neue", Helvetica, sans-serif; }

.nav {
  background-color: #fff;
  display: flex;
  grid-area: nav;
  justify-content: center;
  margin-top: 1.5rem; }
  .nav .menu-item {
    color: #196179;
    padding: 0.1rem 1rem; }

.concepts {
  display: grid;
  flex-flow: column nowrap;
  grid-area: concepts;
  grid-gap: 2rem;
  grid-rows: 1fr 1fr;
  grid-template-columns: repeat(4, 1fr);
  justify-self: center;
  margin-top: 2rem;
  padding: 0; }
  @media screen and (max-width: 1655px) {
    .concepts {
      grid-template-columns: repeat(3, 1fr); } }
  @media screen and (max-width: 1240px) {
    .concepts {
      grid-template-columns: repeat(2, 1fr); } }
  @media screen and (max-width: 820px) {
    .concepts {
      grid-template-columns: 1fr; } }
  .concepts .concept {
    background: white;
    border-radius: 1rem;
    box-shadow: 0 0 12px 3px rgba(0, 0, 0, 0.1);
    display: flex;
    flex-direction: column;
    list-style: none;
    min-height: 20rem;
    padding: 2rem 2rem 1.5rem 2rem;
    width: 20rem; }
    .concepts .concept .image {
      margin-bottom: 1rem;
      width: 20rem; }
      .concepts .concept .image img {
        border: 1px solid lightgrey;
        width: 100%; }
    .concepts .concept .concept-heading {
      align-items: center;
      display: flex; }
      .concepts .concept .concept-heading .title,
      .concepts .concept .concept-heading .title a {
        color: #64bfdf;
        font-size: 1.875rem;
        font-weight: 400;
        margin-right: 0.5rem;
        text-decoration: none; }
    .concepts .concept .author,
    .concepts .concept .author a {
      color: #64bfdf;
      font-weight: 300;
      margin: 1rem 0;
      text-decoration: none; }
    .concepts .concept .tech-stack {
      display: flex;
      flex-flow: row wrap;
      justify-content: flex-start;
      margin-top: auto;
      padding-top: 1rem; }
      .concepts .concept .tech-stack .pill {
        margin: 0.2em 0;
        margin-right: 0.5rem;
        background: #c7c9f4;
        border-radius: 12px;
        color: #fff;
        padding: 0.2rem 0.7rem; }

.footer {
  grid-area: footer; }
