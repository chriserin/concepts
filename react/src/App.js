import React, { Component, useState } from 'react';
import styled from 'styled-components';
import hashrocketHeader from './hashrocket-header.svg';
import conceptsHeader from './concepts-header.svg';
import githubLogo from './github-logo.svg';
import { colors } from './colors.js';
import zip from 'lodash/zip';
import sortBy from 'lodash/sortBy';
import reverse from 'lodash/reverse';
import downCaret from './down-caret.svg';

const ApplicationContainer = styled.div`
  background-color: ${colors.backgroundGrey};
  padding-bottom: 100px;

  a {
    text-decoration: none;
    color: inherit;

    &:hover {
      color: inherit;
    }
  }
`;

const Tagline = styled.div`
  font-family: Sarala;
  font-weight: 700;
  font-size: 19px;
  color: ${colors.dark};
  letter-spacing: 0.9px;
  text-align: left;

  text-decoration: none !important;
`;

const Header = styled.header`
  margin: 24px auto 0px;
  width: 400px;
`;

const HashrocketLogo = styled.div``;

const ConceptsLogo = styled.div`
  margin: 15px 0 8px;
`;

const ConceptsContainer = styled.div`
  display: grid;
  max-width: 1060px;
  margin: 30px auto 0;
  grid-template-columns: repeat(auto-fill, minmax(295px, 1fr));
  grid-gap: 42px;
  transition: all 0.3s ease-in;
`;

const ConceptContainer = styled.div`
  transition: max-height 0.3s ease-in;
  min-height: 250px;
  margin: 10px;
  max-height: ${({ open }) => (open ? '529px' : '250px')};
  max-width: 295px;
  background-color: white;
  position: relative;
  background: #ffffff;
  box-shadow: 1px 1px 7px 1px rgba(199, 199, 199, 0.5);
`;

const Screenshot = styled.div`
  background-image: url('http://hrcpt.online/${props => props.screenshotUrl}');
  background-size:     cover;
  background-repeat:   no-repeat;
  background-position: center center;
  width: 100%;
  height: 166px;
`;

const ImgFilter = styled.div`
  background: rgba(139, 139, 151, 0.44);
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 166px;
`;

const Title = styled.div`
  background-color: #af1f24;
  font-family: Saira;
  font-weight: 700;
  font-size: 23px;
  color: #ffffff;
  letter-spacing: 1.09px;
  text-align: left;
  position: absolute;
  top: 124px;
  left: -12px;
  padding: 0 10px;
`;

const AuthorLine = styled.div`
  margin-top: 10px;
  padding: 0 10px;
  display: flex;
  justify-content: space-between;
`;

const Author = styled.div`
  font-family: Sarala;
  font-weight: 700;
  font-size: 18px;
  color: #232326;
  letter-spacing: 0.85px;
  text-align: left;
  display: inline-block;

  &:hover {
    text-decoration: underline;
  }
`;

const GithubLogo = styled.img`
  display: inline-block;
  width: 25px;
  height: 25px;
  text-align: right;
`;

const TechsLine = styled.div`
  margin-top: 8px;
  padding: 0 10px;
  display: flex;
  justify-content: space-between;
  font-family: Sarala;
  font-weight: 400;
  font-size: 14px;
  color: #000000;
  letter-spacing: 1.35px;
  text-align: left;
`;

const DownCaret = styled.img`
  cursor: pointer;
  transition: all 0.3s ease-in;
  margin-right: 7px;
  transform: ${props => (props.descriptionOpen ? 'rotate(0.5turn);' : 'none')};
`;

const Tech = styled.span`
  cursor: pointer;
  transition: all 0.3s ease-in;
  background-color: ${({ selected }) =>
    selected ? 'rgba(255,41,49,0.32)' : 'inherit'};

  &:first-child {
    margin-left: -3px;
  }

  padding-left: 3px;
  padding-right: 3px;

  &:hover {
    text-decoration: underline;
  }
`;

const Techs = props => {
  const techs = props.techs.map(tech => (
    <Tech
      onClick={() => props.selectTech(tech)}
      selected={props.selectedTech === tech}
      key={tech}
    >
      {tech}
    </Tech>
  ));

  const spans = zip(techs, Array(techs.length - 1).fill(<span>â€¢</span>));

  return <div>{spans}</div>;
};

const DescriptionContainer = styled.div`
  p {
    transition: all 0.3s ease-in;
    padding: 0px 20px 0px 10px;
    opacity: ${props => (!props.open ? 0 : 1)};
    transition-timing-function: ${props =>
      !props.open ? 'ease-in' : 'ease-in'};
    ${props =>
      !props.open
        ? 'max-height: 0; line-height: 5px; font-size: 5px; '
        : 'max-height: 320px; line-height: inherit; font-size: inherit;'};
  }

  font-family: 'EB Garamond';
  font-weight: 400;
  font-size: 14px;
  color: #000000;
  letter-spacing: 1.15px;
  text-align: left;
  line-height: 22px;
`;

const Description = props => {
  return (
    <DescriptionContainer open={props.open}>
      <p>{props.description}</p>
    </DescriptionContainer>
  );
};

const InfoArea = props => {
  return (
    <div>
      <AuthorLine>
        <Author>
          <a href={props.authorUrl}>by {props.author}</a>
        </Author>
        <a href={props.githubUrl}>
          <GithubLogo src={githubLogo} alt={`${props.title} Github`} />
        </a>
      </AuthorLine>
      <TechsLine>
        <Techs
          techs={props.techs}
          selectTech={props.selectTech}
          selectedTech={props.selectedTech}
        />
        <DownCaret
          onClick={props.toggleDescription}
          descriptionOpen={props.descriptionOpen}
          src={downCaret}
        />
      </TechsLine>
      <Description
        open={props.descriptionOpen}
        description={props.description}
      />
    </div>
  );
};

const Concept = props => {
  const [descriptionOpen, setDescriptionOpen] = useState(false);

  const {
    title,
    full_name,
    description,
    languages,
    github_url,
    author_url,
    slug,
  } = props.concept;

  return (
    <ConceptContainer open={descriptionOpen} id={slug}>
      <a href={props.concept.hrcpt_url}>
        <Screenshot screenshotUrl={props.concept.screenshot_url} />
        <ImgFilter />
        <Title>{title}</Title>
      </a>
      <InfoArea
        author={full_name}
        techs={languages}
        githubUrl={github_url}
        description={description}
        authorUrl={author_url}
        toggleDescription={() => {
          setDescriptionOpen(!descriptionOpen);
        }}
        descriptionOpen={descriptionOpen}
        selectTech={props.selectTech}
        selectedTech={props.selectedTech}
        title={title}
      />
    </ConceptContainer>
  );
};

class App extends Component {
  constructor(props) {
    super(props);
    this.state = { selectedTech: '' };
  }

  renderConcepts(concepts, selectTech, selectedTech) {
    return reverse(sortBy(concepts, concept => Date.parse(concept.created_at)))
      .filter(
        concept => !selectedTech || concept.languages.includes(selectedTech)
      )
      .map(concept => (
        <Concept
          key={concept.title}
          concept={concept}
          selectTech={selectTech}
          selectedTech={selectedTech}
        />
      ));
  }

  render() {
    const { concepts } = this.props;

    return (
      <ApplicationContainer>
        <Header>
          <a href="https://hashrocket.com">
            <HashrocketLogo>
              <img src={hashrocketHeader} alt="hashrocket logo" />
            </HashrocketLogo>
          </a>
          <a href="/">
            <ConceptsLogo>
              <img src={conceptsHeader} alt="concepts logo" />
            </ConceptsLogo>
          </a>
          <Tagline>A Gallery For Our Side Projects</Tagline>
        </Header>
        <ConceptsContainer>
          {this.renderConcepts(
            concepts,
            tech =>
              this.setState({
                selectedTech: this.state.selectedTech === tech ? '' : tech,
              }),
            this.state.selectedTech
          )}
        </ConceptsContainer>
      </ApplicationContainer>
    );
  }
}

export default App;
