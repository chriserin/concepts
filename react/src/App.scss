@import url("https://fonts.googleapis.com/css?family=Libre+Franklin:300,400,700");

$pill-color: #c7c9f4;
$pill-text-color: #fff;
$bluey-blue: #64bfdf;
$drawing-blue: darken(rgba(46, 170, 212, 1), 10%);

html,
* {
  font-family: "Libre Franklin", sans-serif;
}

.App {
  display: grid;
  grid-template-rows: repeat(auto-fit, minmax(40px, 1fr));
  grid-template-areas:
    "header"
    "nav"
    "concepts"
    "footer";
}

.App-header {
  background-color: rgba(46, 170, 212, 0.1);
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  align-content: baseline;
  color: white;
  padding: 0.8vh 1vw;
  grid-area: header;
  border-bottom: 1px solid $drawing-blue;

  .hr-container {
    display: flex;
  }

  h2 {
    margin: 0;
    align-self: flex-end;
  }

  h2 a.hr {
    text-decoration: none;
    display: inline-block;
    padding: 0.4rem 0 0rem 1.6rem;
    text-transform: uppercase;
    color: #082736;
    letter-spacing: 0.1em;
    font-weight: 900;
    font-size: 60%;
    line-height: 2;
    color: #ae1f23;
    background: url(https://d15zqjc70bk603.cloudfront.net/assets/brand/hr_logo_icon-81c617c883074d68d44a316868516dc57b3c22eebb0dd3625da85895e554d7fb.svg)
      left 55% no-repeat;
    background-size: 1.2rem auto;
    font-family: Raleway, "Helvetica Neue", Helvetica, sans-serif;
  }
}

.nav {
  background-color: #fff;
  display: flex;
  grid-area: nav;
  justify-content: center;
  margin-top: 1.5rem;

  .menu-item {
    color: darken($drawing-blue, 12%);
    padding: 0.1rem 1rem;
  }
}

.concepts {
  display: grid;
  flex-flow: column nowrap;
  grid-area: concepts;
  grid-gap: 2rem;
  grid-rows: 1fr 1fr;
  grid-template-columns: repeat(4, 1fr);
  justify-self: center;
  margin-top: 2rem;
  padding: 0;

  @media screen and (max-width: 1655px) {
    grid-template-columns: repeat(3, 1fr);
  }

  @media screen and (max-width: 1240px) {
    grid-template-columns: repeat(2, 1fr);
  }

  @media screen and (max-width: 820px) {
    grid-template-columns: 1fr;
  }

  .concept {
    background: white;
    border-radius: 1rem;
    box-shadow: 0 0 12px 3px rgba(0, 0, 0, 0.1);
    display: flex;
    flex-direction: column;
    list-style: none;
    min-height: 20rem;
    padding: 2rem 2rem 1.5rem 2rem;
    width: 20rem;

    .image {
      margin-bottom: 1rem;
      width: 20rem;

      img {
        border: 1px solid lightgrey;
        width: 100%;
      }
    }

    .concept-heading {
      align-items: center;
      display: flex;

      .title,
      .title a {
        color: $bluey-blue;
        font-size: 1.875rem;
        font-weight: 400;
        margin-right: 0.5rem;
        text-decoration: none;
      }
    }

    .author,
    .author a {
      color: $bluey-blue;
      font-weight: 300;
      margin: 1rem 0;
      text-decoration: none;
    }

    .description {
    }

    .tech-stack {
      display: flex;
      margin-top: auto;
      padding-top: 1rem;

      .pill {
        background: $pill-color;
        border-radius: 12px;
        color: $pill-text-color;
        margin-right: 0.5rem;
        padding: 0.2rem 0.7rem;
      }
    }
  }
}

.footer {
  grid-area: footer;
}
