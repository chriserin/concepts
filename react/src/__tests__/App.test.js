import React from 'react';

import renderer from 'react-test-renderer';

import App from '../App';

it('renders the concepts', () => {
  const concepts = [
    {
      title: 'Shapedrawer',
      author: 'chriserin',
      created_at: '2011-06-29T17:11:36Z',
      full_name: 'Chris Erin',
      description: 'This started out as an experiment.',
      languages: ['Ruby', 'JavaScript', 'CoffeeScript'],
      screenshot_url: 'images/shapedrawer.1552425047.png',
      hrcpt_url: 'https://shapedrawer.hrcpt.online',
      author_url: 'https://github.com/chriserin',
      github_url: 'https://github.com/chriserin/ShapeDrawer',
      is_heroku: true,
    },
    {
      title: 'Ceramic Nation',
      author: 'jwworth',
      created_at: '2015-09-11T05:37:54Z',
      full_name: 'Jake Worth',
      description: 'Ceramic Nation is an autogenerated novel.',
      languages: ['Ruby', 'JavaScript'],
      screenshot_url: 'images/ceramic-nation.1552491967.png',
      hrcpt_url: 'https://ceramic-nation.hrcpt.online',
      author_url: 'https://github.com/jwworth',
      github_url: 'https://github.com/jwworth/novel',
      is_heroku: true,
    },
    {
      title: 'Virtuoso',
      author: 'rpmessner',
      created_at: '2017-05-06T02:58:43Z',
      full_name: 'Ryan Messner',
      description:
        'Translate chord names to finger positions on a guitar fretboard',
      languages: ['JavaScript'],
      screenshot_url: 'images/virtuoso.1552490092.png',
      hrcpt_url: 'https://virtuoso.hrcpt.online',
      author_url: 'https://github.com/rpmessner',
      github_url: 'https://github.com/rpmessner/virtuoso',
      is_heroku: false,
    },
  ];

  const tree = renderer.create(<App concepts={concepts} />).toJSON();
  expect(tree).toMatchSnapshot();
});
