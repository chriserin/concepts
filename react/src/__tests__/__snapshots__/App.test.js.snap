// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the concepts 1`] = `
<div
  className="sc-bdVaJa bGXtxv"
>
  <header
    className="sc-htpNat kyUFmG"
  >
    <a
      href="https://hashrocket.com"
    >
      <div
        className="sc-bxivhb bUZGRr"
      >
        <img
          alt="hashrocket logo"
          src="hashrocket-header.svg"
        />
      </div>
    </a>
    <a
      href="/"
    >
      <div
        className="sc-ifAKCX fgFZDe"
      >
        <img
          alt="concepts logo"
          src="concepts-header.svg"
        />
      </div>
    </a>
    <div
      className="sc-bwzfXH eHnsSk"
    >
      A Gallery For Our Side Projects
    </div>
  </header>
  <div
    className="sc-EHOje TylnL"
  >
    <div
      className="sc-bZQynM efeUTq"
      open={false}
    >
      <a
        href="https://virtuoso.hrcpt.online"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="sc-gzVnrw guBrVN"
        />
        <div
          className="sc-htoDjs fhoPnw"
        />
        <div
          className="sc-VigVT jCWGcQ"
        >
          Virtuoso
        </div>
      </a>
      <div>
        <div
          className="sc-fjdhpX buikdd"
        >
          <div
            className="sc-jzJRlG ijxxOu"
          >
            <a
              href="https://github.com/rpmessner"
            >
              by 
              Ryan Messner
            </a>
          </div>
          <a
            href="https://github.com/rpmessner/virtuoso"
          >
            <img
              alt="Virtuoso Github"
              className="sc-cSHVUG fINiSI"
              src="github-logo.svg"
            />
          </a>
        </div>
        <div
          className="sc-kAzzGY cDCBFy"
        >
          <div>
            <span
              className="sc-kgoBCf gltWlu"
              onClick={[Function]}
              selected={false}
            >
              JavaScript
            </span>
          </div>
          <img
            alt="Virtuoso description"
            className="sc-chPdSV kOQaGa"
            onClick={[Function]}
            src="down-caret.svg"
          />
        </div>
      </div>
    </div>
    <div
      className="sc-bZQynM efeUTq"
      open={false}
    >
      <a
        href="https://ceramic-nation.hrcpt.online"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="sc-gzVnrw ekRMNG"
        />
        <div
          className="sc-htoDjs fhoPnw"
        />
        <div
          className="sc-dnqmqq dWyYuZ"
          open={false}
        >
          <div
            className="sc-iwsKbI PZppk"
          >
            <img
              alt="Heroku Logo"
              src="heroku-logo.svg"
            />
            <div>
              We’re waking the Heroku dyno
              <br />
              Please be patient for ~30s
            </div>
            <div
              className="sc-gZMcBi hUnRZf"
            >
              <div
                className="sc-gqjmRU kvYnzV"
              />
            </div>
          </div>
        </div>
        <div
          className="sc-VigVT jCWGcQ"
        >
          Ceramic Nation
        </div>
      </a>
      <div>
        <div
          className="sc-fjdhpX buikdd"
        >
          <div
            className="sc-jzJRlG ijxxOu"
          >
            <a
              href="https://github.com/jwworth"
            >
              by 
              Jake Worth
            </a>
          </div>
          <a
            href="https://github.com/jwworth/novel"
          >
            <img
              alt="Ceramic Nation Github"
              className="sc-cSHVUG fINiSI"
              src="github-logo.svg"
            />
          </a>
        </div>
        <div
          className="sc-kAzzGY cDCBFy"
        >
          <div>
            <span
              className="sc-kgoBCf gltWlu"
              onClick={[Function]}
              selected={false}
            >
              Ruby
            </span>
            <span>
              •
            </span>
            <span
              className="sc-kgoBCf gltWlu"
              onClick={[Function]}
              selected={false}
            >
              JavaScript
            </span>
          </div>
          <img
            alt="Ceramic Nation description"
            className="sc-chPdSV kOQaGa"
            onClick={[Function]}
            src="down-caret.svg"
          />
        </div>
      </div>
    </div>
    <div
      className="sc-bZQynM efeUTq"
      open={false}
    >
      <a
        href="https://shapedrawer.hrcpt.online"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="sc-gzVnrw laNIzT"
        />
        <div
          className="sc-htoDjs fhoPnw"
        />
        <div
          className="sc-dnqmqq dWyYuZ"
          open={false}
        >
          <div
            className="sc-iwsKbI PZppk"
          >
            <img
              alt="Heroku Logo"
              src="heroku-logo.svg"
            />
            <div>
              We’re waking the Heroku dyno
              <br />
              Please be patient for ~30s
            </div>
            <div
              className="sc-gZMcBi hUnRZf"
            >
              <div
                className="sc-gqjmRU kvYnzV"
              />
            </div>
          </div>
        </div>
        <div
          className="sc-VigVT jCWGcQ"
        >
          Shapedrawer
        </div>
      </a>
      <div>
        <div
          className="sc-fjdhpX buikdd"
        >
          <div
            className="sc-jzJRlG ijxxOu"
          >
            <a
              href="https://github.com/chriserin"
            >
              by 
              Chris Erin
            </a>
          </div>
          <a
            href="https://github.com/chriserin/ShapeDrawer"
          >
            <img
              alt="Shapedrawer Github"
              className="sc-cSHVUG fINiSI"
              src="github-logo.svg"
            />
          </a>
        </div>
        <div
          className="sc-kAzzGY cDCBFy"
        >
          <div>
            <span
              className="sc-kgoBCf gltWlu"
              onClick={[Function]}
              selected={false}
            >
              Ruby
            </span>
            <span>
              •
            </span>
            <span
              className="sc-kgoBCf gltWlu"
              onClick={[Function]}
              selected={false}
            >
              JavaScript
            </span>
            <span>
              •
            </span>
            <span
              className="sc-kgoBCf gltWlu"
              onClick={[Function]}
              selected={false}
            >
              CoffeeScript
            </span>
          </div>
          <img
            alt="Shapedrawer description"
            className="sc-chPdSV kOQaGa"
            onClick={[Function]}
            src="down-caret.svg"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
